### PR TITLE
SCMOD-9295: Add healthcheck

### PR DESCRIPTION
**Pull request by dgibson-microfocus**
https://portal.digitalsafe.net/browse/SCMOD-9295

### CODE REVIEW

File: caf-audit/src/main/java/com/hpe/caf/auditing/AuditChannel.java [hunk 1]
```diff
@@ -15,6 +15,9 @@
  */
 package com.hpe.caf.auditing;
 
+import com.hpe.caf.auditing.healthcheck.HealthResult;
+import com.hpe.caf.auditing.healthcheck.HealthStatus;
```
Comment (dermot-hardy) [inline caf-audit/src/main/java/com/hpe/caf/auditing/AuditChannel.java]: Unused import

File: caf-audit-binding-elasticsearch/src/main/java/com/hpe/caf/auditing/elastic/ElasticAuditChannel.java [hunk 2]
```diff
@@ -40,6 +46,22 @@ public AuditEventBuilder createEventBuilder(AuditCoreMetadataProvider coreMetada
...
+    @Override
+    public HealthResult healthCheck()
+    {
+        final ClusterHealthRequest request = new ClusterHealthRequest().waitForYellowStatus();
+        try {
+            final ClusterHealthResponse health = restHighLevelClient.cluster().health(request, RequestOptions.DEFAULT);
+            if(health.getStatus() == ClusterHealthStatus.RED) {
+                return new HealthResult(HealthStatus.UNHEALTHY, "Elasticsearch cluster is unhealthy");
+            }
+        } catch (IOException e) {
```
Comment (dermot-hardy) [inline caf-audit-binding-elasticsearch/src/main/java/com/hpe/caf/auditing/elastic/ElasticAuditChannel.java]:
```suggestion
        } catch (final IOException e) {
```

File: caf-audit-service-admin/pom.xml [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+
+    <dependencies>
+        <dependency>
+            <groupId>com.github.cafaudit</groupId>
+            <artifactId>caf-audit</artifactId>
+        </dependency>
+        <dependency>
+            <groupId>javax.servlet</groupId>
+            <artifactId>javax.servlet-api</artifactId>
+<!--            <version>3.1.0</version>-->
```
Comment (dermot-hardy) [inline caf-audit-service-admin/pom.xml]: Drop this line.

File: caf-audit-service-admin/pom.xml [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+            <scope>runtime</scope>
+        </dependency>
+        <dependency>
+            <groupId>com.fasterxml.jackson.core</groupId>
+            <artifactId>jackson-databind</artifactId>
+        </dependency>
+        <dependency>
+            <groupId>com.fasterxml.jackson.core</groupId>
+            <artifactId>jackson-core</artifactId>
+        </dependency>
```
Comment (dermot-hardy) [inline caf-audit-service-admin/pom.xml:63]: It's no big deal but generally prefer to put the compile-time dependencies first, followed by the runtime dependencies and then the test dependencies.

File: caf-audit-service-admin/src/main/webapp/WEB-INF/web.xml [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+    <display-name>Audit Service Administration Web Application</display-name>
+
+<!--    <servlet>-->
+<!--        <servlet-name>healthcheck</servlet-name>-->
+<!--        <servlet-class>com.hpe.caf.services.audit.server.admin.HealthCheck</servlet-class>-->
+<!--    </servlet>-->
+<!--    <servlet-mapping>-->
+<!--        <servlet-name>healthcheck</servlet-name>-->
+<!--        <url-pattern>/healthcheck</url-pattern>-->
+<!--    </servlet-mapping>-->
```
Comment (dermot-hardy) [inline caf-audit-service-admin/src/main/webapp/WEB-INF/web.xml]: What are all these commented out lines?  Are they needed?  Get rid of them if not.

File: caf-audit-service-admin/src/main/java/com/hpe/caf/services/audit/server/admin/HealthCheck.java [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+
+    @Override
+    protected void doGet(final HttpServletRequest req, final HttpServletResponse resp) throws IOException
+    {
+        HealthResult healthResult;
+        try (final AuditConnection connection = AuditConnectionFactory.createConnection()) {
+            try (final AuditChannel channel = connection.createChannel()) {
+                healthResult = channel.healthCheck();
+                if (healthResult.getStatus() != HealthStatus.HEALTHY) {
+                    resp.setStatus(HttpServletResponse.SC_INTERNAL_SERVER_ERROR);
```
Comment (dermot-hardy) [inline caf-audit-service-admin/src/main/java/com/hpe/caf/services/audit/server/admin/HealthCheck.java]: Double-check that this is the same code that the workers and other services when their health checks fail.

File: caf-audit-service-container/pom.xml [hunk 7]
```diff
@@ -320,6 +324,15 @@
                                                 <unpack>true</unpack>
                                                 <useProjectArtifact>false</useProjectArtifact>
                                             </dependencySet>
+                                            <!-- Including the healthcheck web application. -->
+                                            <dependencySet>
+                                                <outputDirectory>webapps/caf-audit-service-admin</outputDirectory>
+                                                <includes>
+                                                    <include>com.github.cafaudit:caf-audit-service-admin</include>
```
Comment (dermot-hardy) [inline caf-audit-service-container/pom.xml:341]: Shouldn't you also be adding this to the `excludes` block below like the others?

**Review by dermot-hardy — APPROVED** [review]
Approving but there are a few things to do here and I notice that the build is failing so that needs to be sorted out.

File: caf-audit-service-admin/src/main/java/com/hpe/caf/services/audit/server/admin/HealthCheck.java (hunk 6 above)
Comment (dgibson-microfocus) [reply caf-audit-service-admin/src/main/java/com/hpe/caf/services/audit/server/admin/HealthCheck.java]: It's what Job-service currently returns, I'm not sure what workers do.

File: caf-audit-service-container/pom.xml (hunk 7 above)
Comment (dgibson-microfocus) [reply caf-audit-service-container/pom.xml:341]: done

**Comment by dgibson-microfocus** [comment]
@dermot-hardy Yeah I ran into a weird dependency issue. For whatever reason the elastic binding wasn't pulling in the high level rest client despite being a compile dependency. 
I also moved the creation of the `AuditConnection` to the `init` function as it's a relatively expensive function to call every health check.

**Pull request merged**